### PR TITLE
New `multiply_density` option on tallies

### PR DESCRIPTION
**Pull request by paulromano**
Right now, when you ask for reaction rate tallies in OpenMC, they are always computed by multiplying by the atom density (atom/b-cm) of a nuclide present in a material. This effectively gives you macroscopic reaction rates. However, in many situations, you may want a microscopic reaction rate that doesn't depend on the nuclide density (i.e., you just want the microscopic cross section as a multiplier on flux). This PR implements a new `multiply_density` flag on the `Tally` object that allows a user to specify that they do not want reaction rate tallies to be multiplied by the number density of the nuclide within the material. With this capability, you can now tally microscopic reaction rates without a nuclide necessarily being present in a material.

A follow-on PR will refactor other classes within `openmc.mgxs` and `openmc.depletion` to utilize this new functionality.

### CODE REVIEW

**Comment by paulromano** [comment]
Note that I had to update one regression test result for the `tallies` test. Since the test result is hashed, some further explanation is warranted. Here is the diff on the unhashed results:
<details>
<summary>Unhashed test result diff</summary>

```diff
666c666
< 104    22    U235                 events 0.00e+00   0.00e+00
---
> 104    22    U235                 events 1.78e+01   3.77e-01
668c668
< 106    22    U235       inverse-velocity 0.00e+00   0.00e+00
---
> 106    22    U235       inverse-velocity 1.01e-06   4.33e-08
683c683
< 121    22     O16                 events 0.00e+00   0.00e+00
---
> 121    22     O16                 events 1.78e+01   3.77e-01
685c685
< 123    22     O16       inverse-velocity 0.00e+00   0.00e+00
---
> 123    22     O16       inverse-velocity 1.01e-06   4.33e-08
717c717
< 155    23    U235                 events 0.00e+00   0.00e+00
---
> 155    23    U235                 events 3.41e+01   6.14e-01
719c719
< 157    23    U235       inverse-velocity 0.00e+00   0.00e+00
---
> 157    23    U235       inverse-velocity 5.32e-06   1.78e-07
768c768
< 206    60    U235                 events 1.08e+01   2.46e-01
---
> 206    60    U235                 events 7.02e+01   1.24e+00
770c770
< 208    60    U235       inverse-velocity 2.80e-06   1.28e-07
---
> 208    60    U235       inverse-velocity 1.02e-05   3.47e-07
785c785
< 223    60     O16                 events 5.01e+01   8.61e-01
---
> 223    60     O16                 events 7.02e+01   1.24e+00
787c787
< 225    60     O16       inverse-velocity 9.08e-06   3.00e-07
---
> 225    60     O16       inverse-velocity 1.02e-05   3.47e-07
1358c1358
< 104    22    U235                 events 0.00e+00   0.00e+00
---
> 104    22    U235                 events 6.12e-01   7.52e-03
1360c1360
< 106    22    U235       inverse-velocity 0.00e+00   0.00e+00
---
> 106    22    U235       inverse-velocity 1.07e-06   6.29e-08
1375c1375
< 121    22     O16                 events 0.00e+00   0.00e+00
---
> 121    22     O16                 events 6.12e-01   7.52e-03
1377c1377
< 123    22     O16       inverse-velocity 0.00e+00   0.00e+00
---
> 123    22     O16       inverse-velocity 1.07e-06   6.29e-08
1409c1409
< 155    23    U235                 events 0.00e+00   0.00e+00
---
> 155    23    U235                 events 9.37e+00   1.77e-01
1411c1411
< 157    23    U235       inverse-velocity 0.00e+00   0.00e+00
---
> 157    23    U235       inverse-velocity 6.25e-06   1.90e-07
1460c1460
< 206    60    U235                 events 2.76e+00   8.42e-02
---
> 206    60    U235                 events 1.43e+01   2.54e-01
1462c1462
< 208    60    U235       inverse-velocity 2.71e-06   1.39e-07
---
> 208    60    U235       inverse-velocity 1.13e-05   2.79e-07
1477c1477
< 223    60     O16                 events 1.36e+01   2.49e-01
---
> 223    60     O16                 events 1.43e+01   2.54e-01
1479c1479
< 225    60     O16       inverse-velocity 1.00e-05   2.55e-07
---
> 225    60     O16       inverse-velocity 1.13e-05   2.79e-07

```
</details>

As you can see, the `events` and `inverse-velocity` scores are the ones responsible for the differences. If you have a tally with nuclides specified, the current behavior for these two scores is that contributions are made only when the nuclide actually appears in the material (even though these scores don't require nuclide-specific information). With this PR, contributions are made to the `events` and `inverse-velocity` scores regardless of whether the specified nuclide is present in the current material.

**Comment by yrrepy** [comment]
Hi Paul,
this is definitely something that will be very useful.


I'm not sure the comment below belongs here, maybe it better fits in feature requests, but I believe it does involve the same logic blocks: 
I have certain situations where I tally say 2 nuclides of 3 that are present in the material of a tallied cell, and want their total. But 'total' would give me the sum for all 3 nuclides that were in the material.
It would be useful to have a 'subtotal' which would use the listed nuclides.
I had taken a gander and it looked like a fair bit more logic would be needed to get that (beyond my coding depths)

I can of course do the summing in post-processing outside of OpenMC, but having it within permits other facets (plotter visualization, tally triggers, etc.)

File: src/particle.cpp [hunk 1]
```diff
@@ -703,6 +709,29 @@ void Particle::write_restart() const
   } // #pragma omp critical
 }
 
+void Particle::update_neutron_xs(
+  int i_nuclide, int i_grid, int i_sab, double sab_frac, double ncrystal_xs)
+{
+  // Get microscopic cross section cache
+  auto& micro = this->neutron_xs(i_nuclide);
+
+  // If the cache doesn't match, reculcate micro xs
```
Comment (eepeterson) [inline src/particle.cpp:718]: typo in this comment

**Review by eepeterson — CHANGES_REQUESTED** [review]
@paulromano I think this is pretty good to go, just one tiny typo and I think you need to add this attribute to `tallies.rst` too.

**Comment by paulromano** [comment]
Thanks @eepeterson; I've addressed both of your comments.

**Comment by paulromano** [comment]
@yrrepy Thanks for your comment. That's an interesting idea and I see your point about it working properly with all features. It is a bit unrelated to the present feature in this PR though. Could you create a new issue describing this as a feature request so we can track it?

**Review by eepeterson — APPROVED** [review]
LGTM!

**Pull request merged**